service/src/main/java/bio/terra/landingzone/stairway/flight/create/resource/step/CreateAppInsightsStep.java

```diff
@@ -0,0 +1,87 @@
+import bio.terra.landingzone.library.landingzones.definition.factories.ParametersResolver;
+import bio.terra.landingzone.library.landingzones.deployment.LandingZoneTagKeys;
+import bio.terra.landingzone.library.landingzones.deployment.ResourcePurpose;
+import bio.terra.landingzone.stairway.flight.LandingZoneFlightMapKeys;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import com.azure.core.management.exception.ManagementException;
+import com.azure.resourcemanager.applicationinsights.models.ApplicationType;
+import java.util.Map;
+import java.util.UUID;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CreateAppInsightsStep extends BaseResourceCreateStep {
+  private static final Logger logger = LoggerFactory.getLogger(CreateAppInsightsStep.class);
+  public static final String APP_INSIGHT_ID = "APP_INSIGHT_ID";
+
+  public CreateAppInsightsStep(
+      ArmManagers armManagers,
+      ParametersResolver parametersResolver,
+      ResourceNameGenerator resourceNameGenerator) {
+    super(armManagers, parametersResolver, resourceNameGenerator);
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) {
+    var appInsightId = context.getWorkingMap().get(APP_INSIGHT_ID, String.class);
+    try {
+      if (appInsightId != null) {
+        armManagers.applicationInsightsManager().components().deleteById(appInsightId);
+        logger.info("{} resource with id={} deleted.", getResourceType(), appInsightId);
+      }
+    } catch (ManagementException e) {
+      if (StringUtils.equalsIgnoreCase(e.getValue().getCode(), "ResourceNotFound")) {
+        return StepResult.getStepResultSuccess();
+      }
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  protected void createResource(FlightContext context, ArmManagers armManagers) {
+    var landingZoneId =
+        getParameterOrThrow(
+            context.getInputParameters(), LandingZoneFlightMapKeys.LANDING_ZONE_ID, UUID.class);
+    var logAnalyticsWorkspaceId =
+        getParameterOrThrow(
+            context.getWorkingMap(),
+            CreateLogAnalyticsWorkspaceStep.LOG_ANALYTICS_WORKSPACE_ID,
+            String.class);
+
+    var appInsightsName =
+        resourceNameGenerator.nextName(
+            ResourceNameGenerator.MAX_APP_INSIGHTS_COMPONENT_NAME_LENGTH);
+    var appInsight =
+        armManagers
+            .applicationInsightsManager()
+            .components()
+            .define(appInsightsName)
+            .withRegion(getMRGRegionName(context))
+            .withExistingResourceGroup(getMRGName(context))
+            .withKind("java")
+            .withApplicationType(ApplicationType.OTHER)
+            .withWorkspaceResourceId(logAnalyticsWorkspaceId)
+            .withTags(
```

Comment:
Was `withTags` added? Did it happen someplace else before, or was the old code not consistently tagging?

Reply:
Not sure I understand you question. The old code is also tagging app insight resource. It is not as explicit as here. Basically, tagging is happening in CromwellBaseResourcesFactory at 378 (starting point).

Reply:
When I look at the code I don't see it… but maybe it happens as part of some other call?

https://github.com/DataBiosphere/terra-landing-zone-service/blob/47d13216db8d289c5414499fb18adb5f4bac8b62/service/src/main/java/bio/terra/landingzone/library/landingzones/definition/factories/CromwellBaseResourcesFactory.java#L358

Reply:
Yes. It is not straightforward here. But what you can see in CromwellBaseResourcesFactory at line 378 is that this resource will eventually have this tag WLZ-PURPOSE with this value SHARED_RESOURCE and also tag WLZ-ID with landing zone identifier as a value.